server/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java

```diff
@@ -300,24 +324,52 @@ private static ByteSizeValue parse(
         ByteSizeUnit unit,
         final String settingName
     ) {
+        assert unit != BYTES : "Use parseBytes";
         final String s = normalized.substring(0, normalized.length() - suffix.length()).trim();
         try {
             try {
                 return of(Long.parseLong(s), unit);
             } catch (final NumberFormatException e) {
+                BigDecimal decimalValue;
```

Comment:
This logic is pretty complicated and difficult to understand; I was able to understand what's going on but it took me a while and all the nested exceptions...  
Very optional, but maybe you can extract some smaller functions to "name" the steps and make the flow a bit tidier?

Reply:
Ok that's a good idea.